llvm/lib/IRReader/IRReader.cpp

```diff
@@ -91,6 +93,37 @@ std::unique_ptr<Module> llvm::parseIR(MemoryBufferRef Buffer, SMDiagnostic &Err,
                            [](StringRef, StringRef) { return std::nullopt; }));
 }
 
+std::unique_ptr<Module> llvm::parseIR(MemoryBufferRef Buffer, SMDiagnostic &Err,
+                                      LLVMContext &Context,
+                                      ParserCallbacks Callbacks) {
+  llvm::ArrayRef<uint8_t> BufferRef(
+      reinterpret_cast<const uint8_t *>(Buffer.getBufferStart()),
+      reinterpret_cast<const uint8_t *>(Buffer.getBufferEnd()));
+  if (compression::zstd::isAvailable() &&
+      compression::zstd::isZstd(BufferRef)) {
+    llvm::SmallVector<uint8_t> DecompressedBuffer;
+    llvm::Error Error =
+        compression::zstd::decompress(BufferRef, DecompressedBuffer);
+    if (Error) {
+      handleAllErrors(std::move(Error), [&](ErrorInfoBase &EIB) {
+        Err = SMDiagnostic(Buffer.getBufferIdentifier(), SourceMgr::DK_Error,
+                           EIB.message());
+      });
+      return nullptr;
+    }
+
+    std::string Identifier =
+        Buffer.getBufferIdentifier().str() + " (decompressed)";
```

Comment:
What the current practice for "buffer identifiers"? Have you seen other places where they use space and `(` or maybe they just use a snake-case label?

Reply:
I've had a look, but the places where the identifier is not forwarded are sparse, right now I've found a couple of instances where the identifier is `YAML`, `ELF` or `<command line>`. There doesn't seem to be a "prior art" in what to do in this case. I've done it this way because, in my understanding, the identifier is completely cosmetical and so I went with the most "human friendly" identifier

Reply:
Let's go with `-decompressed`, unless you have seen an identifier using spaces.